trie/iterator.go

```diff
@@ -967,51 +964,125 @@ func (it *merkleTreeIterator) seek(path []byte) {
 			if it.err != nil {
 				return
 			}
-		} else {
-			break
+		}
+		break
+	}
+
+	if _, ok := it.stack[len(it.stack)-1].(*merkleTreeIteratorParentNode); ok {
+		it.nextLeaf()
+	}
+
+	if leaf, ok := it.stack[len(it.stack)-1].(*merkleTreeIteratorLeafNode); ok {
+		switch bytes.Compare(path, zk.NewTreePathFromHashBig(common.BytesToHash(leaf.key))) {
+		case -1: // path < leaf path
+		loop1:
+			for it.prevLeaf() {
+				switch bytes.Compare(path, zk.NewTreePathFromHashBig(common.BytesToHash(it.lastNodeAsLeaf().key))) {
+				case -1: // path < leaf path
+					continue
+				case 0: // path == leaf path
+					break loop1
+				case 1:
+					it.nextLeaf()
+					break loop1
+				}
+			}
+		case 0: // leaf path == path
+		case 1: // leaf path < path
+		loop2:
+			for it.nextLeaf() {
+				switch bytes.Compare(path, zk.NewTreePathFromHashBig(common.BytesToHash(it.lastNodeAsLeaf().key))) {
+				case -1: // path < leaf path
+					it.prevLeaf()
+					break loop2
+				case 0: // path == leaf path
+					break loop2
+				case 1:
+					continue
+				}
+			}
+			return
 		}
 	}
 
 	if len(it.path) == 0 {
-		return // root node is not parent node
+		return
 	}
-
 	// When using the Next function, it always moves to the next node.
 	// Therefore, it sets the previous node of the node retrieved by seek to the last visited node.
 	lastIdx := len(it.path) - 1
 	if it.path[lastIdx] == right {
 		it.path[lastIdx] = left
 		it.stack[len(it.stack)-1] = it.resolveNode(it.parentOfLastNode().children[left])
-		for {
-			if parent, ok := it.stack[len(it.stack)-1].(*merkleTreeIteratorParentNode); ok {
-				for _, path := range []byte{right, left} {
-					if child := it.resolveNode(parent.children[path]); child != nil {
-						it.stack = append(it.stack, child)
-						it.path = append(it.path, path)
-						break
-					}
-					if it.err != nil {
-						return
-					}
-				}
-			} else {
-				break
-			}
-		}
+		it.visitBeforeLeafNode()
 	} else {
 		it.path = it.path[:lastIdx]
 		it.stack = it.stack[:len(it.stack)-1]
 	}
 }
 
+func (it *merkleTreeIterator) visitBeforeLeafNode() {
+	for {
+		if parent, ok := it.stack[len(it.stack)-1].(*merkleTreeIteratorParentNode); ok {
+			for _, path := range []byte{right, left} {
+				if child := it.resolveNode(parent.children[path]); child != nil {
+					it.stack = append(it.stack, child)
+					it.path = append(it.path, path)
+					break
+				}
+				if it.err != nil {
+					return
+				}
+			}
+		} else {
+			break
+		}
+	}
+}
+
+func (it *merkleTreeIterator) prevLeaf() bool {
+	if _, ok := it.stack[len(it.stack)-1].(*merkleTreeIteratorLeafNode); ok {
```

Comment:
I think you should check if `len(it.stack) > 0` before this line.

Reply:
Also I think all other places should have this check. like `Hash()`, `NodeBlob()`, `Leaf()`, `parentOfLastNode()`, `lastNodeAsLeaf()`

Reply:
Feedback on stack length 0: https://github.com/kroma-network/go-ethereum/pull/118#discussion_r1855684834